ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/TransactionSimulator.java

```diff
@@ -380,4 +374,87 @@ public Optional<Boolean> doesAddressExist(
 
     return Optional.of(worldState.get(address) != null);
   }
+
+  @FunctionalInterface
+  public interface SimulationGasLimitCalculator {
+    long calculate(final long txGasLimit, final long blockGasLimit);
+
+    static long bounded(final long txGasLimit, final long blockGasLimit, final long rpcGasCap) {
+      final long txGasCap = txGasLimit >= 0 ? txGasLimit : blockGasLimit;
+
+      if (rpcGasCap > 0) {
+        final long simulationGasLimit = Math.min(txGasCap, rpcGasCap);
+        LOG.trace(
+            "Gas limit capped at {} for transaction simulation, tx gas limit={}, block gas limit={}, RPC gas cap={}",
+            simulationGasLimit,
+            txGasLimit,
+            blockGasLimit,
+            rpcGasCap);
+        return simulationGasLimit;
+      }
+      return txGasLimit;
```

Comment:
I'm trying to understand the logic here. If `rpcGasCap == 0` we return txGasLimit and do not consider `txGasCap`? If so, can we move the` final long txGasCap = txGasLimit >= 0 ? txGasLimit : blockGasLimit` to be inside the if?

Also, could you please add a summary to the PR description.